nengo_loihi/loihi_interface.py

```diff
@@ -601,7 +600,7 @@ def _host2chip_spikes(self):
     def _host2chip_errors(self):
         errors = []
         for sender, receiver in self.model.host2chip_senders.items():
-            if isinstance(receiver, PESModulatoryTarget):
+            if hasattr(receiver, "target"):
```

Comment:
Is there a reason behind this change? I feel like a lot of things could have a `target` attribute, but this code is very specific to PES learning (i.e. the errors are passed to a snip that targets them specifically for PES), so I don't see any reason to be general here.

Reply:
Yes, so because the check for `PESModulatoryTarget` got moved from `simulator.py` to `loihi_cx.py` and `loihi_interface.py` there is now a cyclic dependency between `splitter.py` and `loihi_cx.py` because of the `PESModulatoryTarget` check. The cyclic dependency means I needed to check without importing `PESModulatoryTarget` in `loihi_cx.py`, so I made the check the same in `loihi_interface.py` too. I agree that it's not the best check, but `PESModulatoryTarget` is also kind of a weird class that could be replaced by a dict or something else, so I think that will get refactored in the near future anyhow.

Reply:
But yeah, if you prefer a different way to doing the check without importing `PESModulatoryTarget` that's probably fine too (e.g. checking `type(receiver).__name__` or whatever)

Reply:
Typically when I run into these circular dependencies I just put the import at the top of the function. It's not ideal, but as long as one of our unit tests checks it I think that should be fine (as I understand it, the reason for putting imports at the top of a file is so that if they fail, they fail fast. If the imports are intra-repository, I'm less worried about this.) How would you feel about that @tbekolay?

Reply:
I really dislike the imports within functions. IMO the imports at the top of the file are there so the reader can see what this file depends on, i.e., how it interacts with the rest of the codebase (and external libraries). So when another dependency gets introduced in the middle of the file it breaks my mental model of where this file sits in relation to the other files in the repository. The interdependencies between files in a source tree is a big part of my mental model of how the code is organized.

Reply:
I would be happy with moving `PESModulatoryTarget`, `HostSendNode`, `HostReceiveNode`, etc (basically the parts of the splitter after `split_pre_from_host`)... perhaps to `loihi_cx.py`? Or `loihi_api.py`? They'll be moved in a later refactoring anyhow so it's not too critical, but might as well get agreement on it now.

Reply:
I would maybe suggest something like `communication.py`, since all these have to do with host-chip communication. But maybe it's not worth adding another file if it's just going to get changed in refactoring. They don't really fit in either `loihi_cx` or `loihi_api`, though. Where do they go in the refactoring?

Reply:
Or could they go in `builder.py`?

Reply:
They stay in the splitter, but if we're moving them then they'll have to move again in the refactoring, is what I mean. What these nodes are doing should eventually be put into the builder anyhow, so maybe `builder.py`?

Reply:
I hate having conversations on Github, because it really doesn't refresh the messages...

Reply:
Realized that in `builder.py`, we create a circular dependency because it needs `loihi_cx.py`, which needs `PESModulatoryTarget` in `builder.py`. @tbekolay and I talked about this a bit more on Slack, and decided to put them in `loihi_cx.py` temporarily. They'll get moved again in the upcoming refactoring, so just deferring the decision until then.